bee-node/src/node.rs

```diff
@@ -9,49 +9,183 @@
+
+                tangle::tangle().update_snapshot_milestone_index(local_snapshot.metadata().index().into());
+
+                // TODO index 0 ?
+                tangle::tangle().add_solid_entry_point(Hash::zeros(), MilestoneIndex(0));
+                for (hash, index) in local_snapshot.metadata().solid_entry_points() {
+                    tangle::tangle().add_solid_entry_point(*hash, MilestoneIndex(*index));
+                }
+
+                for _seen_milestone in local_snapshot.metadata().seen_milestones() {
+                    // TODO request ?
+                }
+
+                local_snapshot.into_state()
+            }
+            Err(e) => {
+                error!(
+                    "Failed to read snapshot file \"{}\": {:?}.",
+                    self.config.snapshot.local().file_path(),
+                    e
+                );
+                return Err(Error::SnapshotReadError(e));
+            }
+        };
+
+        block_on(Protocol::init(self.config.protocol.clone(), network.clone()));
+
+        // TODO config
+        let (ledger_worker_tx, ledger_worker_rx) = mpsc::channel(1000);
+        let (ledger_worker_shutdown_tx, ledger_worker_shutdown_rx) = oneshot::channel();
+
+        spawn(LedgerWorker::new(snapshot_state.into_balances()).run(ledger_worker_rx, ledger_worker_shutdown_rx));
+
+        info!("Initialized.");
+
+        Ok(Node {
+            config: self.config,
+            network,
+            events: events.fuse(),
+            shutdown,
+            ledger: (ledger_worker_tx, ledger_worker_shutdown_tx),
+            peers: HashMap::new(),
+        })
+    }
+}
 
+/// The main node type.
 pub struct Node {
     config: NodeConfig,
+    // TODO those 2 fields are related; consider bundling them
     network: Network,
-    events: EventSubscriber,
+    events: Fuse<EventSubscriber>,
     shutdown: Shutdown,
-    ledger: Option<(mpsc::Sender<LedgerWorkerEvent>, oneshot::Sender<()>)>,
+    // TODO design proper type `Ledger`
+    ledger: (mpsc::Sender<LedgerWorkerEvent>, oneshot::Sender<()>),
+    // TODO design proper type `PeerList`
     peers: HashMap<EndpointId, (mpsc::Sender<Vec<u8>>, oneshot::Sender<()>)>,
 }
 
 impl Node {
-    pub(crate) fn new(config: NodeConfig, network: Network, events: EventSubscriber, shutdown: Shutdown) -> Self {
-        Self {
-            config,
-            network,
-            events,
-            shutdown,
-            ledger: None,
-            peers: HashMap::new(),
-        }
+    /// Executes node event loop. This method is only executed after the shutdown signal has been received.
+    pub fn run_loop(&mut self) {
+        info!("Running.");
+
+        let mut shutdown = shutdown_listener().fuse();
+
+        block_on(async {
+            loop {
+                select! {
+                    event = self.events.next() => {
+                        if let Some(event) = event {
```

Comment:
I'd separate this whole block out of the select

Reply:
done